### PR TITLE
expr: add MONTH function

### PR DESCRIPTION
**Pull request by chux0519**
## What have you changed? (mandatory)

implement builtin function: month in coprocessor.

## What are the type of the changes? (mandatory)

new feature

## How has this PR been tested? (mandatory)
unittest

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

no

## Does this PR affect tidb-ansible update? (mandatory)
no

## Refer to a related PR or issue link (optional)

#3275

### CODE REVIEW

**Comment by chux0519** [comment]
It seems circle ci has network issue

```
info: downloading component 'rustfmt-preview'
error: component download failed for rustfmt-preview-x86_64-unknown-linux-gnu
info: caused by: could not download file from 'https://static.rust-lang.org/dist/2018-07-18/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz' to '/home/circleci/.rustup/downloads/3e18765f0316079f1787a1ac052082adf50598a139cddafe0420e2e989b6d9fa.partial'
info: caused by: error during download
info: caused by: [35] SSL connect error (OpenSSL SSL_connect: SSL_ERROR_SYSCALL in connection to static.rust-lang.org:443 )
info: backtrace:
```

**Comment by chux0519** [comment]
rebased

**Comment by huachaohuang** [comment]
@solotzg PTAL

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 1]
```diff
@@ -63,6 +63,23 @@ impl ScalarFunc {
         Ok(Some(Cow::Owned(res)))
     }
 
+    #[inline]
+    pub fn month<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<i64>> {
```
Comment (solotzg) [inline src/coprocessor/dag/expr/builtin_time.rs]: Follow the way `fn date` did to handle Error. Thx

Comment (chux0519) [reply src/coprocessor/dag/expr/builtin_time.rs]: modified, and rebased

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 2]
```diff
@@ -203,6 +228,61 @@ mod test {
...
+            let op = Expression::build(&mut ctx, f).unwrap();
+            let got = op.eval(&mut ctx, &[]).unwrap();
+            assert_eq!(got, exp);
+        }
+
+        // test NULL case
+        let input = datum_expr(Datum::Null);
+        let f = scalar_func_expr(ScalarFuncSig::Month, &[input]);
+        let op = Expression::build(&mut ctx, f).unwrap();
+        let got = op.eval(&mut ctx, &[]);
```
[the remark this answers is not in the document]
Comment (chux0519) [reply src/coprocessor/dag/expr/builtin_time.rs]: fixed, `unwrap_err` will be more concise

**Comment by chux0519** [comment]
circle ci fialed when installing tarpaulin, and I do not quite understand the error log, shall we retrigger it?

**Comment by overvenus** [comment]
I have triggered Circle CI, let's see how it goes.

**Comment by overvenus** [comment]
Okay, the latest proc-maco2(v0.4.18) no longer compatible with rust nightly-2018-07-18.

/run-unit-test

**Comment by chux0519** [comment]
That's it. `proc-macro2` updated several hours ago.

**Comment by overvenus** [comment]
PTAL @huachaohuang @breeswish

**Pull request merged**